code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceAllocate.java

```diff
@@ -30,6 +51,43 @@ public InstanceAllocate() {
     }
 
     @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        // Check resourceType for instance and whether the host is a cluster or not.
+        // If resourceType is a cluster (host), do not bother with normal allocation process.
+        final Instance instance = (Instance)state.getResource();
+
+        // check requestedHostId to see whether it's a cluster or not
+        Long requestedHostId = DataAccessor.fields(instance).withKey(InstanceConstants.FIELD_REQUESTED_HOST_ID).as(Long.class);
+        if (requestedHostId != null) {
+            Host potentialCluster = objectManager.loadResource(Host.class, requestedHostId);
+            if (objectManager.isKind(potentialCluster, ClusterConstants.KIND)) {
+                log.info("Instance [{}] requested for a cluster: [{}]", instance, potentialCluster);
+
+                // create instanceHostMap binding with cluster.
+                objectManager.create(InstanceHostMap.class,
+                        INSTANCE_HOST_MAP.HOST_ID, requestedHostId,
+                        INSTANCE_HOST_MAP.INSTANCE_ID, instance.getId());
+
+                // update nic with active subnet
+                Nic nic = nicDao.getPrimaryNic(instance);
+                Network network = objectManager.loadResource(Network.class, nic.getNetworkId());
+                if (network != null) {
```

Comment:
@sonchang ok, should be good then. Below is the explanation why this new class instance.allocate was introduced (for other peoples reference - please correct if anything is missing)

Without Swarm:
=================
1) Instance.create API is called, Instance.create handler logic creates the nics/volumes db entries for the instances and schedules instance.start

2) Instance.start triggers allocators to kick off to allocate the host and the network for the container. Subnet got picked and allocated to a Nic and target host gets picked for the container as a part of this logic.

3) After allocation is done, instance.start handler triggers nic.activate for all container's nics. As a part of nic.activate, vnet gets allocated to the nic.

With Swarm
=================================
Everything above is applied, except the allocation logic. Allocation logic is moved to instance.allocate, and host/subnet are set explicitly on the container/nic in instance.allocate